rfcs/0001-fobnail-architecture-change.md

```diff
@@ -1,143 +1,207 @@
+within a defined amount of time. Otherwise, Fobnail Token will consider the
+Platform untrustworthy.
 
-1. Platform starts attestation by sending signed metadata to Fobnail Token.
-2. Fobnail token verifies metadata and returns Attestation Context ID together
+1. Attester starts attestation by sending signed platform metadata to Fobnail
+   Token.
+2. Fobnail Token verifies metadata and returns Attestation Context ID together
    with PCR selection and nonce.
-3. Platform does TPM quote with its AIK, Fobnail Token provided PCR selection
-   and nonce. Result sent to Fobnail Token, to appropriate Attestation Context.
+3. Attester does TPM quote with its AIK, Fobnail Token provided PCR selection
+   and nonce. Result is sent to Fobnail Token, to appropriate Attestation
+   Context.
 4. Fobnail Token performs Evidence Appraisal and returns trust decision to the
-   Platform.
-5. Fobnail Token unlocks access to keys stored in flash.
+   Attester.
+5. Fobnail Token unlocks access to keys stored in the flash.
+
+![](img/attestation.svg)
 
 # Fobnail Token services
 
-Fobnail services include among others access to protected files and keys. Access
-is granted only to a platform which has passed attestation. Additional, security
-measures could be imposed for finer grained control where each key access would
-have to be granted to each platform.
-
-Secure files can be read, written or deleted. Primary usage for this feature is
-to keep keys which should be possible to read back, e.g. disk encryption keys
-(having Fobnail Token do all crypto on behalf of Platform would result in too
-high performance hit).
-
-Keys, once written can not be retrieved (neither symmetric nor asymmetric). In
-case of symmetric keys Fobnail Token exposes API to encrypt and decrypt data
-using these keys. In case of asymmetric keys Fobnail Token exposes API to
-decrypt or sign data using a private key, and to read public key. Platform is
-responsible for performing cryptographic operations on public key.
-
-Due to Fobnail Token being exposed as a server over (local) network it could be
-accessed by any unprivileged application (just with access to network stack).
-To prevent any application from accessing Fobnail Token services, all requests
-must be signed with AIK which limits access only to application which have
-access to TPM.
+Fobnail services include, among others, access to protected files and keys.
+Access is granted only to a platform that has passed attestation. Additional
+security measures could be imposed for finer-grained control where each key
+access would have to be granted to each Platform.
+
+Secure files can be read, written, or deleted. The primary usage for this
+feature is to keep keys that should be possible to read back, e.g., disk
+encryption keys (having Fobnail Token do all crypto on behalf of Platform would
+result in too high performance hit).
+
+Keys, once written, can not be retrieved (neither symmetric nor asymmetric). In
```

Comment:
@arturkow2000 @krystian-hebel this may lead to issues in production. Question is if we can somehow  rely on KDF to be able to retrieve keys in future or otherwise use Shamir backup with something what is used in cryptocurrency market e.g. [Trezor Shamir Backup](https://trezor.io/shamir/) - I'm aware this would be feature request and subject to additional funding, just let me know if this makes sense.

Reply:
@arturkow2000 

> First, if we were to split key between Fobnail Tokens, Tokens must communicate. They must be plugged into a single computer, and theirs networks must be somehow bridged. 

That's not the point of Shamir Backup for Trezor. Trezor during key generation give you ability to backup phrases which can regenerate private key. Private key never leave Trezor. Those phrases are divided in some set for which you there is subset that triggers recovery. IMHO thise procedure is quite nice I tried it couple times it is long and problematic to some extent, but at least it give you backup that rely on not on electronic device but on physical backup in case of  Trezor Shamir Backup they include [Cryptostell Tubes](https://shop.trezor.io/product/shamir-hodl-pack)

Maybe [this](https://wiki.trezor.io/Shamir_Backup) documentation explain it better.

Reply:
@pietrushnic We need to show recovery phrases to user and we can do this only, over computer, so we must trust the computer to clear these phases when we are done so there is no digital copy. 
Also, I'm wondering how the recovery procedure should look like, we have to enter recovery phrases through computer which we must trust. Couldn't computer by itself regenerate private key?
Also, this backup covers only keys but we may also store other type of data, such as secure files with any contents.

Reply:
@arturkow2000 recovery have to happen on secure/clean room system. This is very sensitive operation so we should assume someone recovers on not compromised machines same thing about provisioning.

I'm not sure what you mean by computer regenerating itself, I though the problem is e.g. loosing of Fobnail Token and loosing access to things which relied on private key which was only in token.

I wonder what other data we should store in Token that would be critical, if encryption keys for disk we should also have backup of that as in case of Heads provisioning.

Reply:
> I'm not sure what you mean by computer regenerating itself

I mean that computer could regenerate private key. Shouldn't the key should be regenerated on a new Fobnail Token. But if we do regeneration in clean room then there is no problem.

Reply:
Most likely it would be on new token.

@arturkow2000 can you answer @dpsmith question? With that I think we would be able to resolve this conversation unless @krystian-hebel has something to add.

Key point from this discussion should be issue in backlog that recovery may be needed.

Reply:
@dpsmith I'm not sure what you mean by releasing keys. Keys are never released in the sense that they may be read. Fobnail may perform cryptographic operations on behalf of platform proved to be trustworthy. No key is accessible before attestation succeeds.

@krystian-hebel Probably we should change that. Currently, we would have to use secure files for storing disk encryption keys to allow reading the key. There is no sense for Fobnail to perform disk decryption as that would be way to slow. We could use a flag (when creating a key) to allow reading of that key.
Doing so will make a clear distinction what is a key and what isn't, and will allow to avoid problems with keys backup.

Reply:
@arturkow2000 what about disk encryption key?

Reply:
@pietrushnic I wrote about it in the comment above - key would be stored as secure file instead. But, I'm thinking about changing this - there would be a flag (set when creating key) to allow reading of the key.

Reply:
@arturkow2000 this is probably discussion not exactly related to this review. We should try to discuss that on matrix. As I said we need some planning for backup, but it does not have to happen here and now.

Reply:
@pietrushnic I opened issue https://github.com/fobnail/fobnail/issues/44